continuationsPlugin/src/main/scala/continuations/ContinuationsPlugin.scala

```diff
@@ -48,6 +50,26 @@ class ContinuationsPhase extends PluginPhase:
     report.logWith("transformBlock")(tree.show)
     tree
 
+  var continuationTraitSym: ClassSymbol = _
+  var continuationObjectSym: TermSymbol = _
+  var safeContinuationClassApplySym: TermSymbol = _
+  var interceptedMethodSym: TermSymbol = _
```

Comment:
💡 I've been creating the transformations as a class `continuations.DefDefTransforms`. I think these could be class parameters and we could avoid the empty vars here if we did it that way.

Reply:
Additionally, since these are compiler-generated, would they be flagged `synthetics`?

Reply:
added the [class](https://github.com/47deg/TBD/pull/52/commits/1a5f9ef8011ed082a480c0da6afa4bae1cc71944#diff-fade65ebf959155389570d62d274801c2f7607f965a273661e5f776636d6393fR19) with parameters but we still need one [var](https://github.com/47deg/TBD/pull/52/commits/1a5f9ef8011ed082a480c0da6afa4bae1cc71944#diff-bfdffcfa12331080aaef50b40eae8ba34ac1bbfa58bb69c99d4ceef4b5607d7cR53). 
not sure if it is a problem but could create the class inside the `transformDefDef`

Reply:
I'm not convinced we actually need those `requireX` lookups outside of the actual `DefDefTransforms` class for this use case. They live in the `Symbols` table of the compiler via loading from the classpath. So inside `DefDefTransforms.transformSuspendContinuation(tree)`, the first thing we could do is pull them in with:

```scala
val ContinuationClazz = requiredClass("continuations.Continuation")
val SafeContinuationsClazz = requiredClass("continuations.SafeContinuation")
val interceptedMethod = requiredPackage("continuations.intrinsics").requiredMethod("intercepted")
...rest of transformSuspendedContinuation(tree)
```
or something similar.

If other methods inside `DefDefTransforms` require the symbols, they can pull them from the `Symbols` object with `Symbols.requiredClass` and `Symbols.requiredPackage`, etc, in the same way. If that gets to be too repetitive, we can look them up in `ContinuationsPlugin.transformDefDef` and then pass them to `DefDefTransforms`'s constructor for reuse (assuming they don't get changed inside our transform).

#### Aside: Sharing mutable state within the phase

There is a `Store` ([Store](https://github.com/lampepfl/dotty/blob/1276ba54bd172c68e4ac733bcef31a4b141b8bc6/compiler/src/dotty/tools/dotc/util/Store.scala#L10)) for generic object storage on `Context`, so to share common values that may change between `transform<TreeType>` methods within the same phase we'd setup the initial state in the `prepare<TreeType>`, add them to the `Context#store` with `Context#addLocation[T](T):Location[T]`([addLocation](https://github.com/lampepfl/dotty/blob/b3f870fdf2f8bedd7d73af0cb80ece5142d2ccec/compiler/src/dotty/tools/dotc/core/Contexts.scala#L660)), and we'd `Promise#tryComplete(Location[T])` to store that in a `Promise[Location[T]]` `val` on the phase. `transform<TreeType>` methods and callees could then look them up with `Context#store#apply(Location[T])`([Store#apply](https://github.com/lampepfl/dotty/blob/1276ba54bd172c68e4ac733bcef31a4b141b8bc6/compiler/src/dotty/tools/dotc/util/Store.scala#L31)), where we get the values from the `Promise[Location[T]]` `val`s on the phase. If we need to mutate the value at `Location[T]`, we do so with `Context#updateStore[T](Location[T], T):Context`([Context#updateStore](https://github.com/lampepfl/dotty/blob/b3f870fdf2f8bedd7d73af0cb80ece5142d2ccec/compiler/src/dotty/tools/dotc/core/Contexts.scala#L672)), which sets the value at `Location[T]` on the context store and returns the context.  We're not doing more than one `transform<TreeType>` transform, yet, so we don't have this problem at this time, but we might need to store contextual values at some point, which is why I included this aside.

A sketch:

```scala
// in ContinuationsPluginPhase
private val myMutableInt = Promise[Int]

override def prepareForApply(tree: Apply)(using Context): Context = 
  val c = summon[Context]
  if(!myMutableInt.isCompleted){
    val _ = myMutableInt.tryComplete(c.addLocation(1))
  }
override def transformApply(tree: Apply)(using Context): Context = 
  new TransformApply(Await.result(myMutableInt.future, Duration.Inf))
  
// . . .
override def transformAssign(tree: Assign)(using Context): Tree =
  new TransformAssign(Await.result(myMutableInt.future, Duration.Inf))

// . . .

// in some method in TransformAssign, which stores the Location[Int] as loc via its constructor
val currentInt = context.store(loc)
val newInt = currentInt + 1
val _ = summon[Context].updateStore(loc, newInt)

// . . .

// in some method in  or TransformApply, which stores the Location[Int] as loc via its constructor
val currentInt = context.store(loc)
val newInt = currentInt + 1
val _ = summon[Context].updateStore(loc, newInt)

// . . .
```
We don't have to use promises and futures, but we need something that can only be completed once to do the setup (because we don't know the order `prepareFor<TypeTree>` will be called in), so it's `Promise` or `CompletableFuture`, I think, for that use case.

This lets us share the mutable `Int` between `transform<TypeTree>` calls and have a consistent state within a compilation unit, without using any mutable vars, and so that we're keeping contextual things in the `Context` where they belong, and allows us to operate on them without having to know the order of application of the transformations. There's also `Context#properties`, but apparently that is slower since `Context#store` is actually just an array, and a `Location[T]` is just a pointer to the index into that array.

Reply:
okay moved the vars inside `transformSuspendNoParametersOneContinuationResume` for now in [0cfb287](https://github.com/47deg/TBD/pull/52/commits/0cfb287f3eed96ca5558eda6d6555533e68838d1). 
as you said if we need them elsewhere we can move them to `transformSuspendContinuation` or higher

>Sharing mutable state within the phase

nice, good to know